### PR TITLE
Feature/608: add SparkPlan rule to fail if there is no data

### PR DESCRIPTION
**Pull request by zzeekk**
see #608

### CODE REVIEW

File: sdl-core/src/main/scala/io/smartdatalake/util/spark/SDLSparkExtension.scala [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+}
+
+case class AssertNotEmptyExec(child: SparkPlan) extends UnaryExecNode {
+
+  // Output has the same attributes as input
+  override def output: Seq[Attribute] = child.output
+
+  // Examine child's data
+  override protected def doExecute() = {
+    val rdd = child.execute()
```
Comment (Geheiner) [inline sdl-core/src/main/scala/io/smartdatalake/util/spark/SDLSparkExtension.scala]: is Spark smart enough to not completely execute the whole query plan a second time for this? It does not look like it

File: sdl-core/src/main/scala/io/smartdatalake/workflow/action/DataFrameActionImpl.scala [hunk 2]
```diff
@@ -352,7 +352,7 @@ abstract class DataFrameActionImpl extends ActionSubFeedsImpl[DataFrameSubFeed]
         subFeed.withMetrics(streamingMetrics.getMainInfos ++ sparkMetrics).asInstanceOf[DataFrameSubFeed]
       case _ =>
         // Auto persist if dataFrame is reused later
-        val preparedSubFeed = if (context.dataFrameReuseStatistics.contains((output.id, subFeed.partitionValues))) {
+        var preparedSubFeed = if (context.dataFrameReuseStatistics.contains((output.id, subFeed.partitionValues))) {
```
Comment (Geheiner) [inline sdl-core/src/main/scala/io/smartdatalake/workflow/action/DataFrameActionImpl.scala]: why does this need to be a var now? I don t see any reassignement

**Review by Geheiner — APPROVED** [review]

File: sdl-core/src/main/scala/io/smartdatalake/util/spark/SDLSparkExtension.scala (hunk 1 above)
Comment (zzeekk) [reply sdl-core/src/main/scala/io/smartdatalake/util/spark/SDLSparkExtension.scala]: Thanks, you're right. This triggers a separate spark job (within the same Query), which is not optimal.
I now found different ways to evaluate the logical plan, without using rdd count or isEmpty.

File: sdl-core/src/main/scala/io/smartdatalake/workflow/action/DataFrameActionImpl.scala (hunk 2 above)
Comment (zzeekk) [reply sdl-core/src/main/scala/io/smartdatalake/workflow/action/DataFrameActionImpl.scala]: fixed

File: sdl-core/src/main/scala/io/smartdatalake/util/spark/SDLSparkExtension.scala (hunk 1 above)
Comment (Geheiner) [reply sdl-core/src/main/scala/io/smartdatalake/util/spark/SDLSparkExtension.scala]: I don t get what you changed compared to before, let's have a look together

**Pull request merged**